pkg/boot/survey/survey.go

```diff
@@ -0,0 +1,367 @@
+	}
+	response, err := root.NewResponse()
+	if err != nil {
+		return nil, err
+	}
+
+	response.SetNamespace(ns)
+
+	if cab, ok := peerstore.GetCertifiedAddrBook(surv.h.Peerstore()); ok {
+		env := cab.GetPeerRecord(surv.h.ID())
+		rec, err := env.Marshal()
+		if err != nil {
+			return nil, err
+		}
+		response.SetEnvelope(rec)
+	}
+	return root.Message().MarshalPacked()
+}
+
+func (surv *Survey) handleMudpResponse(response cpSurvey.SurveyResponse) {
+	var (
+		finder chan peer.AddrInfo
+		rec    peer.PeerRecord
+		err    error
+		ok     bool
+	)
+
+	surv.mu.Lock()
+	defer surv.mu.Unlock()
+
+	ns, err := response.Namespace()
+	if err != nil {
+		return
+	}
+
+	if finder, ok = surv.mustFind[ns]; !ok {
+		return
+	}
+
+	envelope, err := response.Envelope()
+	if err != nil {
+		return
+	}
+
+	if _, err = record.ConsumeTypedEnvelope(envelope, &rec); err != nil {
+		return
+	}
+	select {
+	case finder <- peer.AddrInfo{ID: rec.PeerID, Addrs: rec.Addrs}:
+	default:
+		return
+	}
+}
+
+func (surv *Survey) Advertise(ctx context.Context, ns string, opt ...discovery.Option) (time.Duration, error) {
+	if err := surv.err.Load(); err != nil {
+		return 0, err.(error)
+	}
+
+	surv.mu.Lock()
+	defer surv.mu.Unlock()
+
+	opts, err := surv.options(ns, opt)
+	if err != nil {
+		return 0, err
+	}
+
+	if ttlChan, ok := surv.mustAdvertise[ns]; ok {
+		ttlChan <- opts.Ttl
+	} else {
+		resetTtl := make(chan time.Duration)
+		surv.mustAdvertise[ns] = resetTtl
+		go surv.trackAdvertise(ns, resetTtl, opts.Ttl)
+	}
+	return opts.Ttl, nil
+}
+
+func (surv *Survey) options(ns string, opt []discovery.Option) (opts *discovery.Options, err error) {
+	opts = &discovery.Options{}
+	if err = opts.Apply(opt...); err == nil && opts.Ttl == 0 {
+		opts.Ttl = discTTL
+	}
+
+	return
+}
+
+func (surv *Survey) trackAdvertise(ns string, resetTtl chan time.Duration, ttl time.Duration) {
+	timer := time.NewTimer(ttl)
+	defer timer.Stop()
+	for {
+		select {
+		case <-timer.C:
+			surv.mu.Lock()
+
+			select { // check again TTL after acquiring lock
+			case ttl := <-resetTtl:
+				timer.Reset(ttl)
+				surv.mu.Unlock()
+			default:
+				close(resetTtl)
+				delete(surv.mustAdvertise, ns)
+				surv.mu.Unlock()
+				return
+			}
+		case ttl := <-resetTtl:
+			timer.Reset(ttl)
+		}
+	}
+}
+
+func (surv *Survey) FindPeers(ctx context.Context, ns string, opt ...discovery.Option) (<-chan peer.AddrInfo, error) {
+	if err := surv.err.Load(); err != nil {
+		return nil, err.(error)
+	}
+
+	surv.mu.Lock()
+	defer surv.mu.Unlock()
+
+	var (
+		opts *discovery.Options
+		dist uint8
+		err  error
+	)
+
+	opts, err = surv.options(ns, opt)
+	if err != nil {
+		return nil, err
+	}
+
+	if b, ok := opts.Other["distance"]; ok {
+		dist, err = conv.Uint8(b)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		dist = discDist
+	}
+
+	request, err := surv.buildRequest(ns, dist)
+	if err != nil {
+		return nil, err
+	}
+
+	finder := make(chan peer.AddrInfo, opts.Limit)
+	surv.mustFind[ns] = finder
```

Comment:
#21#discussion_r796845287

> I don't think the channel capacity is an adequate way of limiting responses, since a consumer is popping items off the queue concurrently.